service/matching/matching_engine.go

```diff
@@ -371,60 +371,129 @@ func (e *matchingEngineImpl) getTaskQueuePartitionManager(
+					TaskQueue:     taskQueue.Name(),
+					TaskQueueType: taskQueue.TaskType(),
+					PartitionId:   &taskqueuespb.TaskQueuePartition_NormalPartitionId{NormalPartitionId: int32(partitionId)},
+				},
+			})
+			if err != nil {
+				e.logger.Error("Failed to force load non-root partition after root partition was loaded",
+					tag.WorkflowNamespaceID(nsName.String()),
+					tag.WorkflowTaskQueueName(taskQueue.Name()),
+					tag.WorkflowTaskQueueType(taskQueue.TaskType()),
+					tag.Error(err))
+				return
+			}
+
+			if !resp.WasUnloaded {
+				// Generally we should be avoiding trying to force load partitions which are already loaded.
+				e.metricsHandler.Counter(metrics.ForceLoadedTaskQueuePartitionUnnecessarilyCounter.Name()).Record(1)
+			}
+
+		}()
+	}
+	return nil
+}
+
+// Returns a taskQueuePartitionManager and a bool indicating if the taskQueue was newly created.
+// by attempting to get new range from DB. This does not block for the task queue to be initialized.
+func (e *matchingEngineImpl) createTaskQueuePartitionManager(
 	partition tqid.Partition,
-	create bool,
 	loadCause loadCause,
-) (taskQueuePartitionManager, error) {
+) (taskQueuePartitionManager, bool, error) {
 	key := partition.Key()
-	e.partitionsLock.RLock()
+
+	e.partitionsLock.Lock()
+	defer e.partitionsLock.Unlock()
+
+	// Check taskQueue doesn't already exist
 	pm, ok := e.partitions[key]
-	e.partitionsLock.RUnlock()
-	if !ok {
-		if !create {
-			return nil, nil
-		}
+	if ok {
+		return pm, false, nil
+	}
 
-		// If it gets here, write lock and check again in case a task queue is created between the two locks
-		e.partitionsLock.Lock()
-		pm, ok = e.partitions[key]
-		if !ok {
-			namespaceEntry, err := e.namespaceRegistry.GetNamespaceByID(partition.NamespaceId())
-			if err != nil {
-				e.partitionsLock.Unlock()
-				return nil, err
-			}
-			nsName := namespaceEntry.Name()
-			tqConfig := newTaskQueueConfig(partition.TaskQueue(), e.config, nsName)
-			tqConfig.loadCause = loadCause
-			userDataManager := newUserDataManager(e.taskManager, e.matchingRawClient, partition, tqConfig, e.logger, e.namespaceRegistry)
-			pm, err = newTaskQueuePartitionManager(e, namespaceEntry, partition, tqConfig, userDataManager)
-			if err != nil {
-				e.partitionsLock.Unlock()
-				return nil, err
-			}
-			e.partitions[key] = pm
-		}
-		e.partitionsLock.Unlock()
+	namespaceEntry, err := e.namespaceRegistry.GetNamespaceByID(partition.NamespaceId())
+	if err != nil {
+		return nil, false, err
+	}
 
-		if !ok {
-			pm.Start()
-		}
+	nsName := namespaceEntry.Name()
+	tqConfig := newTaskQueueConfig(partition.TaskQueue(), e.config, nsName)
+	tqConfig.loadCause = loadCause
+	userDataManager := newUserDataManager(e.taskManager, e.matchingRawClient, partition, tqConfig, e.logger, e.namespaceRegistry)
+	pm, err = newTaskQueuePartitionManager(e, namespaceEntry, partition, tqConfig, userDataManager)
+	if err != nil {
+		return nil, false, err
 	}
-	return pm, nil
+
+	pm.Start()
```

Comment:
Start() may be expensive, it should be done after releasing the exclusive lock on partitionsLock

Reply:
Thank you. I really couldn't figure out why it was excluded. I dove through the code looking for attempt to lock the already acquired lock but failed to bring this consideration.